### ISSUE TITLE
Combination of `lia` and `lra`

### ISSUE THREAD

**Issue by maximedenes** [coq/coq#10741]
@damien-pous mentioned during the Coq Workshop the need to sometimes run a combination of `lia` and `lra`. @fajb, do you need if this is feasible today, or could be considered as an extension to what we currently have?

**Comment by damien-pous** [comment]
thanks maxime for relaying the request ; I was thinking of theories like
this one:
http://smtlib.cs.uiowa.edu/theories-Reals_Ints.shtml
(where you have `conversion' functions between integers and reals)

I don't know if it is substantially harder than lia/lra alone.

best,
damien

Le lun. 9 sept. 2019 à 01:30, Maxime Dénès <notifications@github.com> a
écrit :

**Comment by fajb** [comment]
Hi,

I would not know how to make it work right away.

I never looked at the decision procedure but I know it is very similar.
(For the integer case, I currently use a Simplex and cutting planes -- this seems to also work for MILP)

If we would first run a `rify` step, it seems it reduces to a problem for lra where certain variables are obtained by IZR. 
That would be sufficient to discharge goals where the fact that there are integral variables is not relevant.
To be complete, the decision procedure/ proof checker would need to be extended for checking that cutting planes are only obtained using integer variables.

I would say this is doable but not completely straightforward...

A good project for an intern?

**Comment by amahboubi** [comment]
Just to be sure @damien-pous , the smtlib logic you are pointing at is allows _non-linear_ mixed arithmetic. But your question is only about the linear, quantifier free, fragment right?

**Comment by damien-pous** [comment]
Yes, I was targeting linear, quantifier-free, mixed arithmetic (i.e., the
combination of Coq's lia and lra together with knowledge about IZR and
'ceil/floor: R -> Z').


Le lun. 9 sept. 2019 à 15:03, Assia Mahboubi <notifications@github.com> a
écrit :

**Comment by thery** [comment]
I think it is called lira

**Comment by thery** [comment]
Hi @damien-pous . @fajb has something that mixes `lia + lra` , he is looking for examples. Do you know where one can find some benchmarks?

**Comment by damien-pous** [comment]
Hi @thery !

My question was motivated by proofs where I would have appreciated such a combination.
I've put an example online: https://github.com/damien-pous/weierstrass
There I prove Weierstrass' theorem via Bernstein polynomials, and this involves playing with binomial coefficients (on nat) and their injection into real numbers. The proof is finished, and not so long, but there are quite a few places where I believe it could be streamlined even more with a combined lia/lra. 

Any comment welcome on that proof, by the way (I would happily contribute it to whatever library would be a nice fit).
Using mathcomp nats&bigops would make sense, for instance.

Damien

PS: I realise reading this old thread that my answer to @amahboubi is not accurate: I was hoping for some mixed yet linear arithmetic, with quantifiers if possible (i.e., like for lia/lra).

**Comment by damien-pous** [comment]
Hi @thery and @fajb ,

Indeed this is the kind of improvement I'm looking for, thanks!

I've added annotations for other places where I feel rify (or a combined lia+lra tactic) could prove useful at
https://github.com/damien-pous/weierstrass/tree/lira

Damien

**Comment by fajb** [comment]
Hi,

After much sweat, I have a preliminary prototype.
Getting rify working was the not the hardest part. I also wanted `lra` to be able to solve mixed linear integer programs.
```coq
Goal foral (x:R), isZ x -> 2 * x = 1 -> False. Proof. lra. Qed.
```
 Your `Cpion'` lemma can be proved as follows:
```coq
Lemma Cpion': forall n i, (0<n)%nat -> S i / n * C n (S i) = C (pred n) i.
Proof.
  (* TODO: here a mixed lia/lra tactic would help *)
  move=>[/=|n] i. lia.
  move=>_. 
  specialize (Cpion n i).
  simpl.
  tify R. nra.
Qed.
```
If you want to give it a try you need:
- https://github.com/fajb/coq/tree/tify
- https://github.com/fajb/stdlib/tree/rify
- https://github.com/fajb/weierstrass/blob/main/Weierstrass.v

PS: I expect quite a few bugs (correctness/completenes) to be lurking around...
